### PR TITLE
New `rangelim`

### PR DESCRIPTION
**Pull request by numberZero**

### CODE REVIEW

File: src/script/common/c_content.cpp [hunk 1]
```diff
@@ -191,7 +191,7 @@ void read_object_properties(lua_State *L, int index,
 
 	int hp_max = 0;
 	if (getintfield(L, -1, "hp_max", hp_max))
-		prop->hp_max = (s16)rangelim(hp_max, 0, S16_MAX);
+		prop->hp_max = rangelim(hp_max, 0, S16_MAX);
```
Comment (sfan5) [inline src/script/common/c_content.cpp]: shouldn't this be a template too?

File: src/util/numeric.h [hunk 2]
```diff
@@ -34,6 +33,23 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+	if (value < min)
+		return min;
+	if (value > max)
+		return max;
+	return value;
+}
+
+// limits value to range of type T
+template <typename T>
+T rangelim(long value)
```
Comment (SmallJoker) [inline src/util/numeric.h]: How about floating point? `T value` seems more appropriate, maybe even with const ref, as the value is never modified.
Also, `inline`?

File: src/script/common/c_content.cpp (hunk 1 above)
Comment (sfan5) [reply src/script/common/c_content.cpp]: `prop->hp_max = rangelim<s16>(hp_max);`

File: src/util/numeric.h (hunk 2 above)
Comment (numberZero) [reply src/util/numeric.h]: Immediate value is faster. `inline` is sensible but entirely optional, compiler will do that anyway.

File: src/script/common/c_content.cpp (hunk 1 above)
Comment (numberZero) [reply src/script/common/c_content.cpp]: @sfan5 `rangelim` here is the template declared in `numeric.h` You don’t need to use `<>` to use a template *function* (as opposed to template *class*).

**Comment by numberZero** [comment]
Travis status: “Job errored.”

**Comment by SmallJoker** [comment]
I don't see much of a benefit with this PR. Surely a rangelim template would be more C++-ish, but the macro works just fine IMO.

**Comment by ClobberXD** [comment]
Does the new `rangelim` support floats? If not, could it be made to?

**Comment by numberZero** [comment]
> Does the new rangelim support floats?

Both old and new versions should work with floats, and actually with anything linearly ordered (including std::string xD).

> Surely a rangelim template would be more C++-ish, but the macro works just fine IMO.

The macro works well, as long as you pass it plain values. Function calls will be evaluated more than once, and type safety is non-existent. Moreover, the macro is named like a function (all lowercase), making people unaware of these potential problems. But otherwise, the change is more stylistic than functional, typical macro problems are nevertheless unlikely to hit this particular case.

Still, I’d not use a macro for anything that could be implemented as a function. That’s like `goto` which is handy when you lack exceptions (see Linux code), but is hardly ever usable otherwise.

Actually this PR seems to be incomplete. What would be better is to unify all these `core::clamp`, `clamp_u8`, `rangelim`, etc. I think the first `rangelim` should be replaced with `core::clamp` which does the same, while the second should remain as is, it is very handy for e.g. colors (there could be a special function for these, though).

**Comment by paramat** [comment]
No progress for roughly 6 months, and seems to need reworking, closing for now.

**Pull request closed without merging**